dhall-lsp-server/src/Dhall/LSP/Backend/Typing.hs

```diff
@@ -52,14 +55,80 @@ typeAt' pos ctx expr = do
+exprAt pos expr =
   let subExprs = toListOf subExpressions expr
   in case [ (src, e) | (Note src e) <- subExprs, pos `inside` src ] of
     [] -> Nothing
-    ((src, e) : _) -> srcAt pos e <|> Just src
+    ((src,e) : _) -> exprAt pos e <|> Just (Note src e)
+
+
+-- | Find the smallest Src annotation containing the given position.
+srcAt :: Position -> Expr Src a -> Maybe Src
+srcAt pos expr = do e <- exprAt pos expr
+                    case e of
+                      Note src _ -> Just src
+                      _ -> Nothing
+
+
+-- | Given a well-typed expression and a position find the let binder at that
+--   position (if there is one) and return a textual update to the source code
+--   that inserts the type annotation (or replaces the existing one).
+annotateLet :: Position -> Expr Src X -> Maybe (Src, Text)
+annotateLet pos expr = annotateLet' pos empty (splitLets expr)
+
+annotateLet' :: Position -> Context (Expr Src X) -> Expr Src X -> Maybe (Src, Text)
+annotateLet' pos ctx (Note src e@(Let (Binding _ _ a :| []) _))
+  | not $ any (pos `inside`) [ src' | Note src' _ <- toListOf subExpressions e ]
+  = do _A <- rightToMaybe $ typeWithA absurd ctx a
+       let srcAnnot = case getLetAnnot src of
+                        Just x -> x
+                        Nothing -> error "The impossible happened: failed\
+                                         \ to re-parse a Let expression."
+       return (srcAnnot, ": " <> printExpr _A <> " ")
+
+-- binders
+annotateLet' pos ctx (Let (Binding x _ a :| []) e@(Note src _))
+  | pos `inside` src = do
+    _A <- rightToMaybe $ typeWithA absurd ctx a
+    let ctx' = fmap (shift 1 (V x 0)) (insert x _A ctx)
+    annotateLet' pos ctx' e
+annotateLet' pos ctx (Lam x _A b@(Note src _))
+  | pos `inside` src = do
+    let _A' = Dhall.Core.normalize _A
+        ctx' = fmap (shift 1 (V x 0)) (insert x _A' ctx)
+    annotateLet' pos ctx' b
+annotateLet' pos ctx (Pi x _A _B@(Note src _))
+  | pos `inside` src = do
+    let _A' = Dhall.Core.normalize _A
+        ctx' = fmap (shift 1 (V x 0)) (insert x _A' ctx)
+    annotateLet' pos ctx' _B
+
+-- we need to unfold Notes to make progress
+annotateLet' pos ctx (Note _ expr) = do
+  annotateLet' pos ctx expr
+
+-- catch-all
+annotateLet' pos ctx expr =
+  let subExprs = toListOf subExpressions expr
+  in join $ annotateLet' pos ctx <$> listToMaybe [ Note src e | (Note src e) <- subExprs, pos `inside` src ]
+
+
+printExpr :: Pretty.Pretty b => Expr a b -> Text
+printExpr expr = Pretty.renderStrict $ Pretty.layoutCompact (Pretty.unAnnotate (prettyCharacterSet Unicode expr))
+
+
+-- Split all multilets into single lets in an expression
+splitLets :: Expr Src a -> Expr Src a
+splitLets (Note src (Let (b :| (b' : bs)) e)) =
+  splitLets (Note src (Let (b :| []) (Note src' (Let (b' :| bs) e))))
+  where src' = case getLetInner src of
+                 Just x -> x
+                 Nothing -> error "The impossible happened: failed\
```

Comment:
I suggest returning a `Maybe`/`Either` instead of an `error`.  Each of the sites where you use `splitLets` supports gracefully handling the failure case gracefully

Reply:
Hmm that's an interesting design question. Wouldn't we want the entire thing to go up in flames if the "impossible" happens, rather than pretending that nothing happened?

Reply:
The suggestion is not to pretend nothing happens, but to handle the `Nothing` at `splitLets` call site, where you likely have more information available to construct a good error message.

Reply:
Hmm... both call sites return a `Maybe`, where `Nothing` is not necessarily supposed to mean an error has occurred. Another example of this is in
```
annotateLet' pos ctx (Note src e@(Let (Binding _ _ a :| []) _))
  | not $ any (pos `inside`) [ src' | Note src' _ <- toListOf subExpressions e ]
  = do _A <- rightToMaybe $ typeWithA absurd ctx a
       let srcAnnot = case getLetAnnot src of
                        Just x -> x
                        Nothing -> error "The impossible happened: failed\
                                         \ to re-parse a Let expression."
       return (srcAnnot, ": " <> printExpr _A <> " ")
```
where again I think crashing is a better idea than just returning `Nothing`. Basically, failing to parse an expression that successfully parsed earlier can never happen: the branch is in a sense equivalent to `Nothing -> undefined`.

Reply:
Ah I see what you mean now, but I can't think of a more informative error message in the context of `typeAt` or `annotateLet`. Furthermore, I think the type `splitLets :: Expr Src a -> Expr Src a` should be enough to guarantee that the function never fails (assuming the Src annotations come directly from the parser).

Reply:
Looking at it again, I now agree with you after all :) Have a look at the latest commit and tell me what you think.

Reply:
Possibly better, but you're unlikely to convince me that calling `error` is a good idea I'm afraid. If it's a case that `typeAt` and friends are called with data types that are "too big" then that is the exact problem - we need a more accurate data type! Ultimately, this is a data type identical to `Expr`, but with `Let` changed to not accept multiple bindings. Or maybe an `Expr` where the container for `Let` bindings is a parameter, and we can change `NonEmpty` to `Identity`

 However, I'm more of a drive by reviewer, so I hope my comments don't change whether or not this gets merged. I'm a bit of a stickler for correctness.

Reply:
Yeah, I agree with @ocharles that we should strive to make things total from end-to-end when feasible.  That doesn't necessarily mean throwing away the error as a `Nothing`: the `Maybe` type could be changed to an `Either` to preserve the error message.

Or, as @ocharles suggested, we could parametrize the `Expr` type on the number of bindings (i.e. `NonEmpty` vs. `Identity`) so that we could statically guarantee that the expression with split lets has exactly one binding per `Let`

Reply:
Ok, thinking about it some more I think I agree with the majority here ;)

Regarding `Expr`, wouldn't it make much more sense to simply drop lets with multiple binders completely from the AST representation? This would mean we would have to change the pretty printer to omit superfluous `in`s of course...